proposals/list-patterns.md

```diff
@@ -168,153 +114,8 @@ expr.Length is >= 2
-  helper.TryGetStartElement(1, out var element1) && element1 is 1
-}
-```
-
-`collection is { .., 3, 4 }` is lowered to
-```
-@{
-  var helper = new ListPatternHelper(collection, 0, 2);
-
-  helper.Count() >= 2 && // `..` with 2 ending patterns
-  helper.GetEndElement(hatIndex: 2) is 3 && // [^2] is 3
-  helper.GetEndElement(1) is 4 // [^1] is 4
-}
-```
-
-`collection is { 1, 2, .., 3, 4 }` is lowered to
-```
-@{
-  var helper = new ListPatternHelper(collection, 2, 2);
-
-  helper.TryGetStartElement(index: 0, out var element0) && element0 is 1 &&
-  helper.TryGetStartElement(1, out var element1) && element1 is 2 &&
-  helper.Count() >= 4 && // `..` with 2 starting patterns and 2 ending patterns
-  helper.GetEndElement(hatIndex: 2) is 3 &&
-  helper.GetEndElement(1) is 4
-}
-```
-
-The same way that a `Type { name: pattern }` *property_pattern* checks that the input has the expected type and isn't null before using that as receiver for the property checks, so can we have the `{ ..., ... }` *list_pattern* initialize a helper and use that as the pseudo-receiver for element accesses.  
-This should allow merging branches of the patterns DAG, thus avoiding creating multiple enumerators.
-
-Note: async enumerables are out-of-scope. (Confirmed in LDM 4/12/2021)
-Note: sub-patterns are disallowed in list-patterns on enumerables for now despite some desirable uses: `e is { 1, 2, ..[var count] }` (LDM 4/12/2021)
-
 ## Unresolved questions
 
 1. Should we support multi-dimensional arrays? (answer [LDM 2021-05-26]: Not supported. If we want to make a general MD-array focused release, we would want to revisit all the areas they're currently lacking, not just list patterns.)
-1. Should we limit the list-pattern to `IEnumerable` types? Then we could allow `{ 1, 2, ..var x }` (`x` would be an `IEnumerable` we would cook up) (answer [LDM 4/12/2021]: no, we'll disallow sub-pattern in slice pattern on enumerable for now)
-2. Should we try and optimize list-patterns like `{ 1, _, _ }` on a countable enumerable type? We could just check the first enumerated element then check `Length`/`Count`. Can we assume that `Count` agrees with enumerated count?
-3. Should we try to cut the enumeration short for length-patterns on enumerables in some cases? (computing min/max acceptable count and checking partial count against that)
-  What if the enumerable type has some sort of `TryGetNonEnumeratedCount` API?  
-4. Can we detect at runtime that the input type is sliceable, so as to avoid enumeration? .NET 6 may be adding some LINQ methods/extensions that would help. 
+2. Should we accept a general *pattern* following `..` in a *slice_pattern*? (answer [LDM 2021-05-26]: Yes, any pattern is allowed after a slice.)
+3. By this definition, the pattern `{..}` tests for `expr.Length >= 0`. Should we omit such test, assuming `Length` is always non-negative? (answer [LDM 2021-05-26]: `{ .. }` will not emit a Length check)
```

Comment:
Btw, what do you think in terms of formatting?

Question 1: should there be spaces in length-patterns?
    - yes: `[ > 0 ]`, `[ var length ]`
    - no: `[> 0]`, `[var length]`
    - other (user preference, or some other option)

Question 2: should there be a space after the slice `..`?
    - yes: `{​​​​​​​​ 1, 2, .. var rest }​​​​​​​​`
    - no: `{​​​​​​​​ 1, 2, ..var rest }​​​​​​​​`

I assume we'd want spaces in list-patterns: `{ 1, 2, 3 }` rather than `{1, 2, 3}` by default to match object creation. I'll check whether this is driven by a formatting option.

Reply:
I think `[]` should probably just follow indexer access.

For `..`,  we don't have any precedent of this sort of precedence right now. If it was a kind of unary operator, then it was easy: we wanted the operand it to be close to the token. However, since binary operators would be allowed, technically we should use `.. var`, even though at the moment we'd only expect `var` up until maybe user-defined patterns come around. 

I think we should have a style preference here anyways so users don't stuck with the default we choose.

Reply:
For binary operators, we're going to add support for this scenario in the "add parens for clarity" fixer. We'll recommend that parens be added for `..a and b`. That'll be in an upcoming PR.